design/design-generic-references.md

```diff
@@ -0,0 +1,211 @@
+
+The provider pods will resolve the references via managed reconciler, hence they
+will need at least read-only permissions for any resource that's been referenced
+by an instance. Currently, every provider [has read access](https://github.com/crossplane/crossplane/blob/d8f57a8/internal/controller/rbac/provider/roles/roles.go#L66)
+to all of its own kinds, `Secret`s, `ConfigMap`s, `Lease`s and `Event`s in all
+namespaces. In order to reference resources other than the ones listed, users will
+have to create necessary RBAC resources manually to grant the permissions.
+
+### Implementation
+
+The new field will exist in all managed resources and the cross-resource reference
+fields will stay as is and we'll continue adding them wherever possible. They will
+still be the main way to refer to another resource. The generic references will
+help in cases where that's not possible or not implemented yet.
+
+The API struct and the generic resolver will live in Crossplane Runtime and be
+integrated into managed reconciler. Providers will only have to update their
+runtime dependency and regenerate their CRDs to get the feature in.
+
+One of the caveats with following the new conventions is that controller-runtime
+utilities mostly assume that you have Group-Version-Kind of the resource you're
+operating with. In order to use Group-Version-Resource, we will make use of
+[`EquivalentResourceRegistry`](https://github.com/kubernetes/apimachinery/blob/bf1bfd9/pkg/runtime/mapper.go#L44)
+object as a translator between the two, implemented as an alternative to
+[`MustCreateObject`](https://github.com/crossplane/crossplane-runtime/blob/406fe0b/pkg/resource/resource.go#L145).
+After getting a `runtime.Object`, all controller-runtime utilities should be
+available to use.
+
+The resolver will make use of `fieldpath` library we're using for composition to
+get and set the values unknown types and once it is done, an update request will
+be issued. In line with cross-resource references, if the field on the referenced
+object, pointed by `toFieldPath`, already has a value then the resolver will
+skip it, meaning it will resolve only once and users will need to delete that
+value for resolver to set it again.
+
+## Future Considerations
+
+### Transforms
+
+The retrieved value may not be in the form that's useful for the referencer
+resource directly and may need some formatting. If we see a great need for that,
+we can move the transform machinery from Crossplane to Crossplane Runtime and
+reuse it.
+
+### More Limited RBAC
+
+Currently, providers have access to all `Secret`s in the cluster. While this will
+provide a nice UX for generic reference users, it may not be desirable for security
+minded folks since it'd cause giving managed resource permission mean giving access
+to any `Secret` in the cluster. See [#2384](https://github.com/crossplane/crossplane/issues/2384).
```

Comment:
I think we'll need an answer for this before we proceed with this design. If we keep allowing providers to access any secret then implementing this design would mean anyone with access to author any managed resource now has access to indirectly read a value from any secret.

Reply:
By answer, do you mean https://github.com/crossplane/crossplane/issues/2384 should block the implementation of this design?

Reply:
Right - or should at least proceed in parallel. i.e. I don't think we can really treat solving this as a "future consideration".